sqlite: add support for SQLite Session Extension

louwers opened the pull request:
<!--
Before submitting a pull request, please read:

- the CONTRIBUTING guide at https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md
- the commit message formatting guidelines at
  https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

If you believe this PR should be highlighted in the Node.js CHANGELOG
please add the `notable-change` label.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

We were talking about adding support for the [Session Extension](https://www.sqlite.org/sessionintro.html) of SQLite in #53752. This is not a run-time extension but rather a feature built in to the SQLite amalgamation. It can be enabled with a compile flag and is enabled by default on some platforms.

I have never contributed to Node.js before so I will probably need some help to bring this to a conclusion.

---

TODO:

- [x] Make sure sessions are deleted before the database they are attached to is closed
- [ ] Consensus on API
- [x] Documentation
- [x] Allow custom conflict handler
- [x] ~Throw with specific (documented) exception in case of conflict when applying changeset~ since SQLite doesn't consider this to be an error I return `false` when applying the changeset is aborted due to a conflict
- [x] Allow generating a [patchset](https://www.sqlite.org/session/sqlite3session_patchset.html) as well
- [x] Allow specifying table name when creating session (to only track that table)
- [x] Implement Session.close()

---

Example usage:

```js
const database1 = new DatabaseSync(':memory:');
const database2 = new DatabaseSync(':memory:');

database1.exec('CREATE TABLE data(key INTEGER PRIMARY KEY, value TEXT)');
database2.exec('CREATE TABLE data(key INTEGER PRIMARY KEY, value TEXT)');

const session = database1.createSession();

const insert = database1.prepare('INSERT INTO data (key, value) VALUES (?, ?)');
insert.run(1, 'hello');
insert.run(2, 'world');

const changeset = session.changeset();
database2.applyChangeset(changeset);
// Now database2 contains the same data as database1 
```

karimfromjordan commented:
Wouldn't this be solved if users could use their own SQLite build? That's already being considered according to the original issue.

louwers commented:
@karimfromjordan Nope, because you need access to the C API to access this feature.

anonrig commented:
Can you add the appropriate documentation as well?

louwers commented:
I need to make sure sessions are deleted before the database they are attached to is closed.

I think I need the `DatabaseSync` hold onto the `Session`s it creates. Both are `BaseObject`s so I need to look into how to do that.

When a `Session` goes out of scope the corresponding `DatabaseSync` needs to no longer hold onto it. But I also need to make sure that all operations on `Session` objects are no-ops after the database is closed.

louwers commented:
Does anyone know how I format locally?

louwers commented:
@anonrig I have added documentation, most of the functionality and fixed the lint issues. Could you trigger another CI run?

I want to add support for generating a patchset and then this PR is ready for review.

I have some ideas for future improvements, i.e. more fine-grained conflict handling, but they can be added in a backward-compatible manner.

TheOneTheOnlyJJ commented on test/parallel/test-sqlite.js:
```diff
@@ -867,11 +867,31 @@ suite('session extension', () => {
+    database1.prepare(insertSql).run(1, 'hello');
+    database2.prepare(insertSql).run(1, 'world');
+    const result = database2.applyChangeset(session.changeset());
+    t.assert.strictEqual(result, false, "expected applyChangeset to return false, indicating an conflict due to an abort");
```
"... indicating **an** conflict ...." typo should be changed to "... indicating **a** conflict ..."

louwers replied:
Thanks for taking a look. This typo is no longer present in the most recent commit.

TheOneTheOnlyJJ commented:
This is excellent work that many projects will surely be based on. I am curious about the eventual wrapping of the entire Sessions C API (mixed & mashed into higher-level Node functions). Does this PR leave space for that? There are a few more Sessions API functions that would be of great use in Node.

louwers commented on src/node_sqlite.cc:
```diff
@@ -653,6 +837,79 @@ BaseObjectPtr<StatementSync> StatementSync::Create(Environment* env,
+  return tmpl;
+}
+
+void Session::MemoryInfo(MemoryTracker* tracker) const {}
```
I am not sure how to implement these.

louwers replied:
@jasnell Do you know if anything needs to happen here?

jasnell replied:
For now I'd say it is ok to leave this as is.

louwers commented:
@TheOneTheOnlyJJ Thanks! I agree that exposing more functionality of the Session Extension in Node.js would be great. Here are some ideas:

- Wrappers for [`sqlite3changeset_invert()`](https://www.sqlite.org/session/sqlite3changeset_invert.html) (creates a changeset that undoes the changes in another changeset) and [`sqlite3changeset_concat()`](https://www.sqlite.org/session/sqlite3changeset_concat.html) (combines changesets).
- Allowing the contents of a changeset to be read. You probably want to use a [`Generator`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Generator) for this.
- Closing a session. Right now a session is closed when it is garbage collected or when the corresponding database it closed. It will probably be helpful to be able to manually close session.
- Change the `onConflict` handler to take a function as well. You would be able to inspect the conflicting change and decide on the action (omit, replace, abort) for each conflicting change.
- A wrapper for [`sqlite3session_diff `](https://www.sqlite.org/session/sqlite3changegroup_add_strm.html).
- Utilize [streaming versions](https://www.sqlite.org/session/sqlite3changegroup_add_strm.html) of the Session API. These are helpful for reading or applying changesets that do not fit in memory.

Here is the full API for future reference: https://www.sqlite.org/session/funclist.html

Right now I first want to gather some feedback, but I'd be interested continue working on this, either in a follow-up PR or in this one. Since the SQLite integration is still under active development I think working incrementally is the preferred approach.

TheOneTheOnlyJJ commented:
Good point, here's my view on them:

> * Wrappers for [`sqlite3changeset_invert()`](https://www.sqlite.org/session/sqlite3changeset_invert.html) (creates a changeset that undoes the changes in another changeset) and [`sqlite3changeset_concat()`](https://www.sqlite.org/session/sqlite3changeset_concat.html) (combines changesets).

Yes, I was about to suggest these. There's also [`sqlite3session_isempty()`](https://www.sqlite.org/session/sqlite3session_isempty.html) and [`sqlite3session_memory_used()`](https://www.sqlite.org/session/sqlite3session_memory_used.html) which are simpler (compared to the rest of the API) and could be wrapped more easily.

> * Allow closing a session. Right now a session is closed when it is garbage collected or when the corresponding database is closed. It is helpful to be able to manually close session in some cases.

This should be a high priority right now, as long-running Node processes (like server back-ends) that would use Sessions will have their memory slowly but surely filled up with no way of freeing it without closing the database connection. Applications could run for months or even longer without closing a database connection, so not being able to close Sessions and free their memory would deter developers from using them. The database should, of course, still keep a list of all active sessions and delete them when it is closed, just in case not all Sessions were manually closed. From my point of view, this function should be wrapped as soon as possible and be included in this PR. Not having it surely decreases the chances of getting this merged, and it would be a shame to lose out on this.

> * Allowing the contents of a changeset to be read. You probably want to reuturn a [`Generator`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Generator) for this so you don't need to read the entire changeset it in memory at once.
> * Change the `onConflict` handler to take a function as well. You would be able to inspect the conflicting change and decide on the action (omit, replace, abort) for each conflicting change.
> * A wrapper for [`sqlite3session_diff()`](https://www.sqlite.org/session/sqlite3changegroup_add_strm.html).

Very important functionalities, but again, they're not vital for achieving functional basic interaction with the extension. These 3 points could be bundled in a follow-up PR that handles the finer-grained data management side of the API. The [`changegroup`](https://www.sqlite.org/session/sqlite3changegroup_new.html) could be exposed as an object with its API functions bound to it as methods.

> * Utilize [streaming versions](https://www.sqlite.org/session/sqlite3changegroup_add_strm.html) of the Session API (you would need another API for Node.js as well). These are helpful for reading or applying changesets that do not fit in memory.

I've noticed this in the documentation and it's definitely a part of the API that would be crucial in large apps with large databases. But again, for the current goal of incrementally proposing a stable API that exposes the API of the extension, this is not required right now. It would also require wrapping [`sqlite3session_config()`](https://www.sqlite.org/session/sqlite3session_config.html), which allows configuring `SQLITE_SESSION_CONFIG_STRMSIZE`. This could be handled after the 3 points above get sorted out.

Also, before I wrap up...

I noticed that you proposed the `createSession()` method as a wrapper around both `sqlite3session_create()` and `sqlite3session_attach()`, combined. Looking at the API, I'm trying to figure out if there's any potential functionality or use case where exposing these functions separately in Node would make sense. The only potential use case I see is creating Sessions in one part of the application and then attaching them somewhere else. Maybe this would be useful if using a global Session that gets created at app start and attached later in the app life cycle? But, given that a Session cannot be unattached, this is likely not going to be useful in any way, so the separation of the functions is not needed. What's you input on this, do you see any potential use cases for the splitting of these 2 functions?

louwers commented:
> This should be a high priority right now

Agreed. I will still add this one as part of this PR.

> What's you input on this, do you see any potential use cases for the splitting of these 2 functions?

I don't see any benefits. Furthermore, I like the [philosophy](https://github.com/WiseLibs/better-sqlite3/blob/master/docs/contribution.md#2-simplicity) of better-sqlite3:

> better-sqlite3's public API must be as simple as possible. Rather than calling 3 functions in a specific order, it's simpler for users to call a single function. Rather than providing many similar functions for doing similar things (e.g., "convenience functions"), there should just be one function that is already convenient by design. Sane defaults should be applied when possible. A function's minimal call signature should be as small as possible, with progressively complex customization available when needed. Function names should only be as long as necessary to convey their purpose. For any new feature, it should be easy to showcase code examples that is are so simple that they are self-explanatory.

In this vein I also oped to default to aborting on conflict, instead of requiring an explicit conflict handler to be passed like in the API of SQLite.

TheOneTheOnlyJJ commented:
> Agreed. I will still add this one as part of this PR.

Yes, this would be ideal. For further updates other than this, I believe it's best to wait for the final form of the `node:sqlite` API itself.

> I don't see any benefits. Furthermore, I like the [philosophy](https://github.com/WiseLibs/better-sqlite3/blob/master/docs/contribution.md#2-simplicity) of better-sqlite3:

I was checking if any additional functionality could be squeezed out by separating the functions. Given that there is none, your decision is a sensible one and I support it.

> In this vein I also oped to default to aborting on conflict, instead of requiring an explicit conflict handler to be passed like in the API of SQLite.

This is also the best sane default option, as every developer should customize this behavior according to their use case.

I'm not experienced in the internals of Node, so my contribution here can only amount to discussing the API design and smaller implementation details.

I guess the next step now (after implementing the Session `close` wrapper) is to wait for a maintainer to review this?

RedYetiDev commented:
FWIW `sqlite` is now a valid subsystem

TheOneTheOnlyJJ commented:
The Session close wrapper should be added to the TODO list at the top of the Issue.

louwers commented:
@TheOneTheOnlyJJ Added TODO and the code. 👍

louwers commented:
I merged main back in and made use of the new `DatabaseSync::IsOpen` method.

Could @anonrig or @cjihrig perhaps trigger a CI run?

louwers commented:
@cjihrig There was a merge conflict on CI for some reason. I merged main back in again, could you trigger another run? 🙂

louwers commented:
Not sure why CI is reporting a merge conflict:

```
17:44:52 HEAD is now at d5316e8426 Merge remote-tracking branch 'origin/main' into session-extension
17:44:52 + git status
17:44:52 HEAD detached at d5316e8426
17:44:52 Untracked files:
17:44:52   (use "git add <file>..." to include in what will be committed)
17:44:52 	env.properties
17:44:52 
17:44:52 nothing added to commit but untracked files present (use "git add" to track)
17:44:52 + git rev-parse HEAD
17:44:52 d5316e8426b89634a76a9926cfa887488b9d4523
17:44:52 + git rev-parse origin/main
17:44:52 e272cfbf05adc8d8687aad9ede5660ab99b360c6
17:44:52 + [ -n origin/main ]
17:44:52 + git rev-parse origin/main
17:44:52 + REBASE_ONTO=e272cfbf05adc8d8687aad9ede5660ab99b360c6
17:44:52 + git rebase --committer-date-is-author-date e272cfbf05adc8d8687aad9ede5660ab99b360c6
17:44:53 Rebasing (1/33)
17:44:53 Auto-merging src/node_sqlite.cc
17:44:53 Auto-merging src/node_sqlite.h
17:44:53 CONFLICT (content): Merge conflict in src/node_sqlite.h
17:44:53 Auto-merging test/parallel/test-sqlite.js
17:44:53 error: could not apply bc951f733b... src, test: support for SQLite Session Extension
```

louwers commented:
@cjihrig CI was passing, but I just had to rebase again.

Would love to get some feedback on this PR. 🙏

cjihrig commented:
> Would love to get some feedback on this PR.

I'll give the code a look. I think we need a `@nodejs/sqlite` team that can be pinged on PRs like this. I don't want to become a blocker (I don't currently have a lot of free time for Node, and the time I do have leading up to v23 will be largely spent on the test runner). I think it would also be good for getting more opinions on what to include in the API.

louwers commented on src/node_sqlite.cc:
```diff
@@ -239,6 +256,175 @@ void DatabaseSync::Exec(const FunctionCallbackInfo<Value>& args) {
+}
+
+void DatabaseSync::ApplyChangeset(const FunctionCallbackInfo<Value>& args) {
+  conflictCallback = nullptr;
+  filterCallback = nullptr;
```
I could use an anonymous namespace instead, but SQLite needs function pointers.

cjihrig replied:
I guess it's fine since we can't have two calls running at the same time. We can revisit in the future if it becomes an issue.

jasnell commented on src/node_sqlite.h:
```diff
@@ -84,6 +93,34 @@ class StatementSync : public BaseObject {
+  template <Sqlite3ChangesetGenFunc sqliteChangesetFunc>
+  static void Changeset(const v8::FunctionCallbackInfo<v8::Value>& args);
+  static void Close(const v8::FunctionCallbackInfo<v8::Value>& args);
+  void MemoryInfo(MemoryTracker* tracker) const override;
+  static v8::Local<v8::FunctionTemplate> GetConstructorTemplate(
+      Environment* env);
+  static BaseObjectPtr<Session> Create(Environment* env,
+                                       BaseObjectWeakPtr<DatabaseSync> database,
+                                       sqlite3_session* session);
+
```
Just minor nit... let's keep the `MemoryInfo` related stuff together.

```suggestion
  static v8::Local<v8::FunctionTemplate> GetConstructorTemplate(
      Environment* env);
  static BaseObjectPtr<Session> Create(Environment* env,
                                       BaseObjectWeakPtr<DatabaseSync> database,
                                       sqlite3_session* session);

  void MemoryInfo(MemoryTracker* tracker) const override;
```

jasnell commented on src/node_sqlite.cc:
```diff
@@ -94,9 +100,19 @@ DatabaseSync::DatabaseSync(Environment* env,
+  for (auto* session : sessions_) {
+    sqlite3session_delete(session);
+  }
+  sessions_.clear();
```
If the `sqlite3_session*` can be wrapped into an `std::unique_ptr` with a custom deleter then we don't need to iterate over them and delete before clearing. Would be a bit safer.

louwers replied:
I think that would go against the expectations of a `std::unique_ptr`, since a `Session` object also have that same `sqlite3_session*`.

jasnell replied:
Ah, I missed that they are shared. Can these use `std::shared_ptr` in some way?

louwers replied:
A `std::shared_ptr` calls a deleter when all owners have let go. But in this case we need to make sure that whenever any owner wants to let go (when the session or database is closed), the session is deleted. Maybe this can be refactored with a weak_ptr at some point, but the Statement class in this file uses a similar pattern.

jasnell commented on src/node_sqlite.cc:
```diff
@@ -239,6 +256,175 @@ void DatabaseSync::Exec(const FunctionCallbackInfo<Value>& args) {
+        String::NewFromUtf8(env->isolate(), "table", NewStringType::kNormal)
+            .ToLocalChecked();
+    if (options->HasOwnProperty(env->context(), table_key).FromJust()) {
+      Local<Value> table_value =
+          options->Get(env->context(), table_key).ToLocalChecked();
```
Nit: it is generally preferable to avoid use of `ToLocalChecked()` if possible. Something like

```suggestion
      Local<Value> table_value;
      if (!options->Get(env->context(), table_key).ToLocal(&table_value)) {
       return;
      }
```

jasnell commented on src/node_sqlite.cc:
```diff
@@ -239,6 +256,175 @@ void DatabaseSync::Exec(const FunctionCallbackInfo<Value>& args) {
+    Local<Object> options = args[1].As<Object>();
+
+    Local<String> conflictKey =
+        String::NewFromUtf8(
+            env->isolate(), "onConflict", NewStringType::kNormal)
+            .ToLocalChecked();
```
Likely better to add this to the set of strings available via `Environment`... e.g. `env->onconflict_string()`

louwers replied:
Could you give me a hint how to do that?

jasnell replied:
Take a look at env_properties.h

jasnell commented on src/node_sqlite.cc:
```diff
@@ -239,6 +256,175 @@ void DatabaseSync::Exec(const FunctionCallbackInfo<Value>& args) {
+        String::NewFromUtf8(
+            env->isolate(), "onConflict", NewStringType::kNormal)
+            .ToLocalChecked();
+    if (options->HasOwnProperty(env->context(), conflictKey).FromJust()) {
+      Local<Value> conflictValue =
+          options->Get(env->context(), conflictKey).ToLocalChecked();
```
Nit: why not just unconditionally try to get the value and ignore it if the value returned is `undefined`? Would save the extra double property lookup here.

louwers replied:
Resolved, but please have a look.

jasnell commented on src/node_sqlite.cc:
```diff
@@ -239,6 +256,175 @@ void DatabaseSync::Exec(const FunctionCallbackInfo<Value>& args) {
+
+    Local<String> filterKey =
+        String::NewFromUtf8(env->isolate(), "filter", NewStringType::kNormal)
+            .ToLocalChecked();
```
Nit: move to `Environment`... e.g. `env->filter_string()`

TheOneTheOnlyJJ commented:
> * Utilize [streaming versions](https://www.sqlite.org/session/sqlite3changegroup_add_strm.html) of the Session API (you would need another API for Node.js as well). These are helpful for reading or applying changesets that do not fit in memory.

Just for reference in the future, this part of the API could get heavy inspiration from #54213, once it gets finalised, approved & merged.

TheOneTheOnlyJJ commented:
Is there any update on this?

louwers commented:
@TheOneTheOnlyJJ Working on the PR comments. I intend to get this merged!

halvardssm commented:
I am sorry to intrude into this conversation, but I have been working on creating a standardized interface for SQL based databases for JS (runtime agnostic) for a little while. Since I see that the API for node:sqlite has not yet reached consensus, I was hoping to join the discussion and see if an API that works for multiple database drivers can be collaborated on. If this would be of interest, I would be happy to join a conversation, or you can join the conversation over at the RFC. Let me know if there is anything I can clarify. Looking forward to see this being natively supported!

RFC: https://github.com/halvardssm/stdext/pull/6

louwers commented on doc/api/sqlite.md:
```diff
@@ -153,6 +153,74 @@ added: v22.5.0
+    * `SQLITE_CHANGESET_ABORT`: abort on conflict and roll back databsase (default).
+* Returns: {boolean} Whether the changeset was applied succesfully without being aborted.
+
+An exception is thrown if the database is not
```
I used the same wording as elsewhere in the document. For consistency we should either keep it the same here or update it everywhere.

louwers replied:
But we just use plain `Error`s at this point, so I don't think it's worth it to write that everywhere.

louwers commented on doc/api/sqlite.md:
```diff
@@ -153,6 +153,74 @@ added: v22.5.0
+  * `filter` {Function} Optional function. Skips changes when a truthy value is returned from it.
+    Takes the name of the table that a change targets as first argument. When this option is not
+    provided all changes are attempted.
+  * `onConflict` {number} Determines how conflicts are handled. When provided, must be one of the values below:
+    * `SQLITE_CHANGESET_OMIT`: conflicting changes are omitted.
+    * `SQLITE_CHANGESET_REPLACE`: conflicting changes replace existing values.
+    * `SQLITE_CHANGESET_ABORT`: abort on conflict and roll back databsase (default).
```
The default is the one which has **(default)**.

But I'll try to make it clearer.

louwers commented on test/parallel/test-sqlite.js:
```diff
@@ -97,3 +102,353 @@ test('PRAGMAs are supported', (t) => {
+    };
+
+    const databaseFrom = createDatabase();
+    const session = databaseFrom.createSession();
```
I'm not so sure, since the Session object is not exported.

louwers commented on doc/api/sqlite.md:
```diff
@@ -153,6 +153,74 @@ added: v22.5.0
+  * `onConflict` {number} Determines how conflicts are handled. When provided, must be one of the values below:
+    * `SQLITE_CHANGESET_OMIT`: conflicting changes are omitted.
+    * `SQLITE_CHANGESET_REPLACE`: conflicting changes replace existing values.
+    * `SQLITE_CHANGESET_ABORT`: abort on conflict and roll back databsase (default).
```
Done.

louwers commented on doc/api/sqlite.md:
```diff
@@ -153,6 +153,74 @@ added: v22.5.0
+* `options` {Object} An optional object used to configure the session.
+  * `table` {string} When provided, only changes to this table are tracked by the created session.
+    By default, changes to all tables are tracked.
+  * `db` {string} Name of the database to track. Default: `'main'`.
```
I added an explanation.

louwers commented on src/node_sqlite.h:
```diff
@@ -3,6 +3,8 @@
 
 #if defined(NODE_WANT_INTERNALS) && NODE_WANT_INTERNALS
 
+#define SQLITE_ENABLE_SESSION
```
I also added it to `node.gyp` so that it also applies to this header file.

louwers commented on test/parallel/test-sqlite.js:
```diff
@@ -97,3 +102,353 @@ test('PRAGMAs are supported', (t) => {
+    t.assert.strictEqual(database1.prepare(select2).all().length, 2);  // data1 should have values in database1
+  });
+
+  const prepareConflict = () => {
```
Good idea!

louwers commented on test/parallel/test-sqlite.js:
```diff
@@ -3,7 +3,12 @@
+  DatabaseSync,
+  SQLITE_CHANGESET_OMIT,
+  SQLITE_CHANGESET_REPLACE,
+  SQLITE_CHANGESET_ABORT
```
My 2 cents is that auto imports work better with this approach.

And these constants will be very unique, so there's not really a reason to namespace them further.

louwers commented:
Thanks for your reviews! Took me a while to get back to this.

I have addressed most comments and left a few questions. Please have another look if you have the chance.

If someone could kick off CI that would be great.

aduh95 commented:
```
src/node_sqlite.cc:383:  Tab found; better to use spaces  [whitespace/tab] [1]
src/node_sqlite.cc:475:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
Done processing src/node_sqlite.cc
Total errors found: 2
```

louwers commented:
@cjihrig @aduh95 Pushed another commit for formatting.

louwers commented on src/node_sqlite.cc:
```diff
@@ -211,6 +222,172 @@ void DatabaseSync::Exec(const FunctionCallbackInfo<Value>& args) {
+
+      if (table_value->IsString()) {
+        String::Utf8Value str(env->isolate(), table_value);
+        table = std::string(*str);
```
Thanks for the review! Could you elaborate? It doesn't compile.

I changed it to `table = *str;` to simplify it slightly.

louwers commented:
Happy to see that CI is green! I made one more small fixes to address a PR comment from @jasnell to include strings in the environment.

Let me know if there's anything else. 🙏

louwers commented:
@TheOneTheOnlyJJ Thank you, I hope so. Sure, I have [some ideas](https://github.com/nodejs/node/pull/54181#issuecomment-2267610396) for follow-up work. What would be your priority? I will create an issue this week.

TheOneTheOnlyJJ commented:
> @TheOneTheOnlyJJ Thank you, I hope so. Sure, I have [some ideas](https://github.com/nodejs/node/pull/54181#issuecomment-2267610396) for follow-up work. What would be your priority? I will create an issue this week.

Your ideas are solid. It makes sense to pursue them as you mentioned.

Maybe the rest of the simpler features could be implemented first, as they would expose more API area of the extension 
to developers ASAP, while the more complex features would come later, one by one, as they get implemented.
Among these simpler features are functions like [`sqlite3session_isempty`](https://sqlite.org/session/sqlite3session_isempty.html) and [`sqlite3session_memory_used`](https://sqlite.org/session/sqlite3session_memory_used.html).

But as an even greater priority, I believe we should create a table that maps the C API of the extension to the Node API implemented up to date directly, with "Not implemented" noted where there's no Node equivalent for the native C API yet (functions, constants, etc.).

I believe this would:
- make it easier for maintainers to keep track of which parts of the API are covered and which are not
- make it easier for willing contributors to find what they could contribute with
- potentially make the process of onboarding new contributors & maintainers easier (I am convinced that this PR took some effort on your part, and having this bring more contributors in would take some pressure off)

Also, if included in the official documentation, this table would:
- make it easier to navigate through the Node API for developers who are more accustomed to the C API
- spare developers the frustration of finding out that an API part they need is not yet covered by the Node API

ReillyBrogan commented:
I arrived here after trying to figure out why we were getting an `undefined symbol` error after updating to nodejs 22.12.0 on Solus. 

Many Linux distributions currently build nodejs with `--shared-sqlite` so that it uses the system libsqlite, however most of them don't build the session extension currently which will cause the nodejs build to succeed* however it will fail upon launch with a `undefined symbol: sqlite3session_delete` error. The nodejs build should catch an incompatible system libsqlite and error out with an appropriate message so that this can be caught during build-time.

* I'm not sure how it's succeeding, it _should_ fail during linking if the symbols aren't present in the libsqlite3.so.

cjihrig commented:
@louwers can you update the implementation to return errors if Node is built without the session extension?

louwers commented:
@cjihrig I will see what I can do.

louwers commented:
@cjihrig The docs state that:

  It is the responsibility of any distribution
  shipping with these options to:
  
  * ensure that the shared dependencies available at runtime
    match what is expected by the Node.js binary. A
    mismatch may result in crashes or unexpected behavior.
  * fully test that Node.js operates as expected with the
    external dependencies. There may be little or no test coverage
    within the Node.js project CI for these non-default options.

I'm going to say this works as intended...

richardlau commented:
> 
> * I'm not sure how it's succeeding, it _should_ fail during linking if the symbols aren't present in the libsqlite3.so.

On `main` I get link failures if building with `--shared-sqlite`.

louwers commented:
You need to pass `SQLITE_ENABLE_SESSION` and `SQLITE_ENABLE_PREUPDATE_HOOK` as defines when compiling SQLite.

ReillyBrogan commented:
> You need to pass `SQLITE_ENABLE_SESSION` and `SQLITE_ENABLE_PREUPDATE_HOOK` as defines when compiling SQLite.

Thank you, I'd already figured this out though. Perhaps this should be added to the release notes for the most recent 22.x and 23.x releases? Just as an informational note for those distributions using a shared sqlite3 (unsupported as it may be).



> > * I'm not sure how it's succeeding, it _should_ fail during linking if the symbols aren't present in the libsqlite3.so.
> 
> On `main` I get link failures if building with `--shared-sqlite`.

Ah, I see the issue on our side now. Since nodejs has such a heavy linking phase we were using mold to shave some time off the build, but apparently mold doesn't throw errors on undefined symbols... (https://github.com/rui314/mold/issues/659). I get appropriate errors with ld and lld.

louwers commented:
@ReillyBrogan I have made PR to update the changelog. https://github.com/nodejs/node/pull/56318
